v1.9 backports 2021-01-14

errordeveloper opened the pull request:
* #14501 -- pkg/datapath: fix arp ping handling (@aanm)
 * #14452 -- AKS: Clean up azure-vnet state on Cilium agent start (@ti-mo)
 * #14565 -- clustermesh: Ignore symlink files on fsnotify events (@tgraf)
 * #14577 -- Fix clustermesh-apiserver dependencies on pkg/option (@tgraf)
 * #14596 -- bpf: fix misconfigured nat to 0.0.0.0 on !masquerade config (@borkmann)
 * #14516 -- Pr/jrajahalme/remove cidr fix (@jrajahalme)
 * #14576 -- ci: check if gke cluster has a nodepool before reserving it (@nebril)
 * #14591 -- pkg/node: fix concurrent access of entry node (@aanm)
 * #14557 -- k8s: Decouple CNP embedding inside CCNP (@christarazi)
 * #14521 -- hubble relay: various logging improvements (@kaworu)
 * #14616 -- nodeinit: use azure0 presence as a condition for flushing ebtables & neigh (@ti-mo)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 14501 14452 14565 14577 14596 14516 14576 14591 14557 14521  14616; do contrib/backporting/set-labels.py $pr done 1.9; done
```

ti-mo committed:
nodeinit: remove nodeinit.expectAzureVnet Helm option and behaviour

[ upstream commit 868eb5d99abace525a9280277fc0291f5efb9b3f ]

Commit 0b70117b9 ("AKS: Fix dynamic reconfiguration of bridge mode") and
preceding commits would modify an internal state file of the azure-vnet CNI
plugin. This is potentially dangerous, because:

- azure-vnet uses a lockfile to coordinate access to this state file between
  independent invocations.
- The format of the state file is not controlled by us and subject to change.
- How this state file is used can change between versions of azure-vnet.

After investigating some reports of connectivity issues, it turns out that:

- Modifying the state file directly isn't necessary. azure-vnet obeys the mode
  it's given in its CNI configuration, regardless of the mode recorded in the
  state file.
- Changing the mode directly _in the state file_ from `bridge` to `transparent`
  causes azure-vnet to forget cleaning up static ARP entries for deleted Pods
  that were previously created in bridge mode before Cilium was deployed.
  This can cause addresses of Pods previously scheduled on the node to become
  unroutable on said node.
- Waiting for the state file to appear causes a deadlock on newly-created nodes
  in a cluster where Cilium is already running, for example, when scaling out
  a node pool. azure-vnet only creates the state file when it gets its first
  CNI `ADD` event, which will never happen with the CNI configurations Cilium
  installs.

Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

jrajahalme committed:
ip: Fix RemoveCIDR edge condition

[ upstream commit 6ad5a22b8644fc23fb5dec2291a80ffce8d8657e ]

A CIDR should be able to be removed from itself.

Add test cases that would fail without the fix:

$ go test

----------------------------------------------------------------------
FAIL: ip_test.go:184: IPTestSuite.TestRemoveCIDRsEdgeCase

ip_test.go:190:
    s.testIPNetsEqual(allowedCIDRs, expectedCIDRs, c)
ip_test.go:83:
    c.Assert(created, HasLen, len(expected))
... obtained []*net.IPNet = []*net.IPNet(nil)
... n int = 1

----------------------------------------------------------------------
FAIL: ip_test.go:194: IPTestSuite.TestRemoveCIDRsEdgeCase2

ip_test.go:200:
    s.testIPNetsEqual(allowedCIDRs, expectedCIDRs, c)
ip_test.go:83:
    c.Assert(created, HasLen, len(expected))
... obtained []*net.IPNet = []*net.IPNet(nil)
... n int = 1

----------------------------------------------------------------------
FAIL: ip_test.go:176: IPTestSuite.TestRemoveSameCIDR

ip_test.go:180:
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"allow CIDR prefix must be a superset of remove CIDR prefix"} ("allow CIDR prefix must be a superset of remove CIDR prefix")

OOPS: 14 passed, 3 FAILED

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

ti-mo committed:
nodeinit: Azure IPAM - remove azure-vnet.json, flush ebtables, ip neighbors

[ upstream commit 9ae10c5dfdabde5581d3034caf6b78d1c47f6a6c ]

This aims to address #14233 ("Pods scheduled before Cilium don't get state
cleaned up") until a better solution can be implemented for AKS, ideally by
gaining support for deploying Cilium directly by Azure.

Currently, when Cilium is deployed with Azure IPAM enabled, all workload Pod
connectivity is interrupted until the Pod is re-scheduled.

Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

ti-mo committed:
Documentation: warn for disconnected Pods when deploying Azure IPAM

[ upstream commit 21d2ebc2e85a785afdb1e8cb88ec57ac6217ab25 ]

Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

aanm committed:
pkg/datapath: fix arp ping handling

[ upstream commit 0ea5ef96876419545dcb6f4664ec19a93813be39 ]

Due some underlying concurrency issues, the ARP requests were dropped by
the kernel ("arp_error_report") as the neighbor lookup fails. This
caused the unit test to fail while waiting for an ARP reply, for which a
ARP request was never sent. To avoid this issue, a time.Sleep of 1
second is added in the unit test.

Fixes: c5ed9aeb217e ("node: Add unit test for node arpinging")
Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

kaworu committed:
certloader/fswatcher: fsnotify log refactoring

[ upstream commit e3815d528ed20b63b90933fd268bb28a76e45411 ]

Only log fsnotify error on close when there is one, and at the warn
level rather than debug.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

kaworu committed:
hubble/relay: raise the log level of messages

[ upstream commit e8b5838a5839380ed906cc83a159ec441923c594 ]

Before this patch, debugging connectivity issue from Relay to Hubble
could be challenging without the debug flag on.

This patch make it so messages are logged at the Info level.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

kaworu committed:
hubble/relay: improve mTLS configuration mismatch logging

[ upstream commit 39ae444bbfe45a0321a813268f79f6e22087e6e4 ]

Before this patch, when Relay was configured with mTLS and Hubble was
configured without mTLS (or conversely), the same error message would be
shown ("context deadline exceeded").

We can detect both cases from the Relay side before attempting to
connect easily: we know whether Relay  was configured with or without
mTLS, and the Peer Service provide the Hubble TLS information.

This patch detect a mismatch in mTLS configuration and log a helpful
error message hinting at the issue.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

christarazi committed:
k8s: Decouple CNP embedding inside CCNP

[ upstream commit 8e9d511e850cc6825c4e7268f511a54141781330 ]

Previously, embedding CNP inside the CCNP object was meant to simplify
the code as a CCNP is an identical structure to CNP.

However, due to a bug reported by a user
(https://github.com/cilium/cilium/issues/14526), we've uncovered that
CCNP validation in the preflight command was completely broken. Upon
further investigation, it was due to an unmarshalling issue with the
CCNP object (that embeds a CNP).

Throughout the code, we have instances where we must treat the CCNP
object with special care to avoid this unmarshalling issue (see
b5c04eec48df3049f78694747059e3f43fc3d847 and
c813a15041dd8996a686a218ffae61d74b4ef137). In the above reported issue,
yet again we can apply the same workarounds as the aforementioned
commits do.

However, we should avoid papering over this again with the same
workaround, because in the future someone may not be aware of this
gotcha when they make a new change. It is better to simply move away
from embedding CNP inside CCNP and deal with the code duplication that
results from it, than to rely on the mental energy and worst of all,
waste time trying to debug very unintuitive errors within K8s
infrastructure code. This commit moves away from embedding CNP inside
CCNP.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

christarazi committed:
validator: Expand tests to include full CNP & CCNP

[ upstream commit 3510d24fbe1a42834dd8875299c8b64517baef41 ]

Due to https://github.com/cilium/cilium/issues/14526, we should test
against full YAMLs representing CNPs and CCNPs for more realistic
results.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

christarazi committed:
docs: Update docs referencing CCNP

[ upstream commit ff02aafd4421d9df436e7fef816699efd8a5c475 ]

Following the previous commit, update the references in the docs for
CCNP.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

christarazi committed:
docs: Update docs referencing Go struct of CNP

[ upstream commit 79f98123665c5baaa990fd81d98b684bba95fcf3 ]

Bring this up-to-date with what's currently in the codebase.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

tgraf committed:
clustermesh: Ignore symlink files on fsnotify events

[ upstream commit f03450706a4b3ca4fcfd17def021a74114aac7f5 ]

Kubernetes secrets are mapped into the pod using symlinks. The initial
scan was already correctly ignoring symlinks but the fsnotify events
have not been. This has resulted in invalid cluster configurations being
added:

```
ClusterMesh:            0/3 clusters ready, 0 global-services
   cluster2: not-ready, 0 nodes, 0 identities, 0 services, 0 failures (last: never)
   └  Waiting for initial connection to be established
   ..2021_01_08_21_11_57.892158678: not-ready, 0 nodes, 0 identities, 0 services, 0 failures (last: never)
   └  Waiting for initial connection to be established
   ..data: not-ready, 0 nodes, 0 identities, 0 services, 0 failures (last: never)
   └  Waiting for initial connection to be established
```

Fixes: 076b0188b98 ("Inter cluster connectivity (ClusterMesh)")

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

kaworu committed:
hubble/relay: log messages refactoring

[ upstream commit 882e37f784728cbc21bceec73d041ab7cfa3ec02 ]

Leverage logrus ability to create structured logs.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

nebril committed:
ci: check if gke cluster has a nodepool before reserving it

[ upstream commit 9606dd802fe4c0c4c9ab9bffc78c037541092030 ]

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

aanm committed:
pkg/node: fix concurrent access of entry node

[ upstream commit a544c516d10f7c1d83988c717bdd257bdb0ce8ea ]

Fixes the following data race:

```
Read at 0x00c00079f200 by goroutine 43:
  github.com/cilium/cilium/pkg/node/manager.(*Manager).StartNeighborRefresh.func1.1()
      /go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:581 +0xc4

Previous write at 0x00c00079f200 by goroutine 166:
  github.com/cilium/cilium/pkg/node/manager.(*Manager).NodeUpdated()
      /go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:414 +0x824
  github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).ciliumNodeInit.func2()
      /go/src/github.com/cilium/cilium/pkg/k8s/watchers/cilium_node.go:71 +0x2d4
  k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:238 +0xa5
  k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnUpdate()
      <autogenerated>:1 +0x2e
  github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1()
      /go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:114 +0x389
  k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:507 +0x50d
  k8s.io/client-go/tools/cache.(*controller).processLoop()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:183 +0x83
  k8s.io/client-go/tools/cache.(*controller).processLoop-fm()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:181 +0x44
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x75
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xba
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x114
  k8s.io/apimachinery/pkg/util/wait.Until()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x507
  k8s.io/client-go/tools/cache.(*controller).Run()
      /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:154 +0x4a9

Goroutine 43 (running) created at:
  github.com/cilium/cilium/pkg/node/manager.(*Manager).StartNeighborRefresh.func1()
      /go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:578 +0x368
  github.com/cilium/cilium/pkg/controller.(*Controller).runController()
      /go/src/github.com/cilium/cilium/pkg/controller/controller.go:208 +0xd10

Goroutine 166 (running) created at:
  github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).ciliumNodeInit()
      /go/src/github.com/cilium/cilium/pkg/k8s/watchers/cilium_node.go:101 +0x184
```

Fixes: 5ec4d51f980f ("daemon, node: refresh neighbor by sending arping periodically")
Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

tgraf committed:
clustermesh-apiserver: Support K8sServiceProxyName

[ upstream commit 7babfa5c883f9caf030d5b41ecd16fc75f46d887 ]

The K8sServiceProxyName affects the subset of services Cilium should
implement and thus must be respected by ClusterMesh as well.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

tgraf committed:
clustermesh-apiserver: Correctly set the cluster name

[ upstream commit eb2b649b2859c17904017e17969f52f82043055d ]

The package pkg/service/store is consumed by the clustermesh-apiserver.
It depends on pkg/option internally. Due to not setting up
option.Config.ClusterName correctly, default values have been used.
Refactor pkg/service/store to no longer depend on pkg/option and require
users to pass in the required configuration settings via an interface.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

borkmann committed:
bpf: fix misconfigured nat to 0.0.0.0 on !masquerade config

[ upstream commit 47cddb511112d078c4e71494e60aba63829d356f ]

Thomas reported:

  When I do Pod to Pod I see this on the interface of the VM:

  21:30:11.022617 IP 0.0.0.0.53336 > 10.240.0.22.8080: Flags [S], seq 2523807568, win 64240,
    options [mss 1460,sackOK,TS val 207396274 ecr 0,nop,wscale 7], length 0

  When I do Host to Pod I see this on the interface of the same VM:

  21:30:02.275063 IP 10.240.0.35.46236 > 10.240.0.22.8080: Flags [S], seq 1718112282, win 64240,
    options [mss 1460,sackOK,TS val 2667439682 ecr 0,nop,wscale 7], length 0

The NAT table indeed had a bunch of entries that would cause the 0.0.0.0 translation:

  [...]
  TCP OUT 10.240.0.16:60250 -> 10.0.222.20:443 XLATE_SRC 0.0.0.0:60250 Created=17376sec HostLocal=0
  TCP OUT 10.240.0.16:46342 -> 10.240.0.49:9153 XLATE_SRC 0.0.0.0:46342 Created=17376sec HostLocal=0
  [...]

From the config, it turns out that the following was set:

  [...]
  enable-bpf-masquerade: "true"
  masquerade: "false"
  [...]

And looking at the headers, ENABLE_MASQUERADE was enabled. This in fact meant that
in snat_v4_needed() we'd look up the endpoint based on the src IPv4 and given neither
ENABLE_IP_MASQ_AGENT nor ENCAP_IFINDEX was set, IPV4_MASQUERADE is set for the NAT
address. However, the latter will be 0.0.0.0 in this case since IPV4_MASQUERADE is
not defined from template side given it depends on ...

  if option.Config.EnableIPv4Masquerade && option.Config.EnableBPFMasquerade && bpfMasqIPv4Addrs != nil {
      if option.Config.EnableIPv4 {
         ipv4 := bpfMasqIPv4Addrs[ifName]
         opts["IPV4_MASQUERADE"] = byteorder.HostSliceToNetwork(ipv4, reflect.Uint32).(uint32)
      }
  }

... of which option.Config.EnableIPv4Masquerade is false. Yet ENABLE_MASQUERADE otoh
is defined in a more relaxed manner, that is option.Config.EnableBPFMasquerade &&
option.Config.EnableIPv4, both of which were true in this case. Fix it by requiring
option.Config.EnableIPv4Masquerade instead of option.Config.EnableIPv4 for the latter.

The commit got uncovered by a0ec2ad991c3 ("datapath: Decouple IPV4_MASQUERADE from
IPV4_NODEPORT"), but the original commit introducing the issue is 0962c0298491
("datapath: Enable BPF MASQ for veth mode in IPv4").

Fixes: 0962c0298491 ("datapath: Enable BPF MASQ for veth mode in IPv4")
Related: 5462cedd3e5d ("datapath: Do not set exclusion CIDR if IPv4 is disabled")
Related: a0ec2ad991c3 ("datapath: Decouple IPV4_MASQUERADE from IPV4_NODEPORT")
Reported-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

borkmann committed:
cilium: auto-disable bpf masq if v4 masq is not enabled

[ upstream commit d678513f602ba93d2195bebbf7a3261e673b9cb3 ]

Given the former is a more specific subflag of the latter for the agent,
we should auto-disable it and log an info message in the agent. This needs
to be adapted when we also implement v6 BPF-based masq.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

aanm reviewed (approved):
LGTM for my changes

kaworu reviewed (approved):
LGTM for my patches, thanks!

ti-mo committed:
nodeinit: use azure0 presence as a condition for flushing ebtables & neigh

[ upstream commit 51f174f1f582f6813e1aa624e1117608703427d8 ]

With https://github.com/cilium/cilium/pull/14192, the assumption that
10-azure-vnet.conflist is still present when nodeinit runs is invalidated.

Split AKS node-init in two stages; one to ensure azure-vnet is configured
in bridge mode and another that only runs when configured with Azure IPAM
that flushes ebtables & neigh when azure0 is present.

Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

christarazi reviewed (approved):
LGTM for my commits
